### PR TITLE
Add ckeditor templates in sonata_formatter_type and sonata_simple_formatter_type

### PR DESCRIPTION
**Pull request by ternel**
I am targetting this branch, because it's a patch that add possibility without any BC break.

No issue about CKeditor's template.

## Changelog

```markdown
### Added
- Added the possibility to use `templates` defined in your `ivory_ckeditor.yml` in the `sonata_formatter_type` and `sonata_simple_formatter_type`
```

## Subject

If you are using `templates` from CKeditor / IvoryCKeditorBundle, they are now automatically included in the template by `sonata_formatter_type` and `sonata_simple_formatter_type`.

### CODE REVIEW

File: Form/Type/FormatterType.php [hunk 1]
```diff
@@ -48,19 +49,26 @@ class FormatterType extends AbstractType
     protected $pluginManager;
 
     /**
+     * @var TemplateManagerInterface
+     */
+    protected $templateManager;
```
Comment (greg0ire) [inline Form/Type/FormatterType.php]: Should be private until proven needed in child classes.

File: Tests/Form/Type/FormatterTypeTest.php [hunk 2]
```diff
@@ -384,6 +388,42 @@ public function testBuildViewWithDefaultConfigAndPluginManager()
...
             'ckeditor_basepath' => '',
             'ckeditor_plugins' => array(),
+            'ckeditor_templates' => array(),
+            'ckeditor_toolbar_icons' => array(),
+        ));
+
+        $this->assertSame($view->vars['ckeditor_configuration'], $defaultConfigValues);
+    }
+
+    public function testBuildViewWithDefaultConfigAndPluginManagerAndTemplateManager()
```
Comment (OskarStark) [inline Tests/Form/Type/FormatterTypeTest.php]: could you please add a test where ckeditor_templates is not empty?

**Comment by OskarStark** [comment]
Great contribution @ternel 👍

File: Form/Type/FormatterType.php [hunk 3]
```diff
@@ -167,9 +175,14 @@ public function buildView(FormView $view, FormInterface $form, array $options)
...
         }
 
+        if (null !== $this->templateManager && $this->templateManager->hasTemplates()) {
+            $options['ckeditor_templates'] = $this->templateManager->getTemplates();
+        }
+
         $view->vars['ckeditor_configuration'] = $ckeditorConfiguration;
         $view->vars['ckeditor_basepath'] = $options['ckeditor_basepath'];
         $view->vars['ckeditor_plugins'] = $options['ckeditor_plugins'];
+        $view->vars['ckeditor_templates'] = $options['ckeditor_templates'];
```
Comment (greg0ire) [inline Form/Type/FormatterType.php]: Is `$options['ckeditor_templates']` guaranteed to be defined…

File: Form/Type/FormatterType.php [hunk 4]
```diff
@@ -222,6 +235,7 @@ public function configureOptions(OptionsResolver $resolver)
             'ckeditor_basepath' => 'bundles/sonataformatter/vendor/ckeditor',
             'ckeditor_context' => null,
             'ckeditor_plugins' => array(),
+            'ckeditor_templates' => array(),
```
Comment (greg0ire) [inline Form/Type/FormatterType.php]: … by this?

Comment (OskarStark) [inline Form/Type/FormatterType.php]: yes it is, i thought about this, too

Comment (ternel) [inline Form/Type/FormatterType.php]: yep, this is the purpose of setDefaults()

File: Form/Type/FormatterType.php [hunk 5]
```diff
@@ -48,19 +49,26 @@ class FormatterType extends AbstractType
     protected $pluginManager;
 
     /**
+     * @var TemplateManagerInterface
+     */
+    private $templateManager;
```
Comment (ternel) [inline Form/Type/FormatterType.php]: private ;)

File: Tests/Form/Type/FormatterTypeTest.php [hunk 6]
```diff
@@ -429,4 +429,54 @@ public function testBuildViewWithDefaultConfigAndPluginManagerAndTemplateManager
 
         $this->assertSame($view->vars['ckeditor_configuration'], $defaultConfigValues);
     }
+
+    public function testBuildViewWithDefaultConfigAndPluginManagerAndTemplateManagerAndWithTemplates()
```
Comment (greg0ire) [inline Tests/Form/Type/FormatterTypeTest.php]: `With` appears twice, is it deliberate?

File: Tests/Form/Type/FormatterTypeTest.php [hunk 7]
```diff
@@ -384,9 +388,95 @@ public function testBuildViewWithDefaultConfigAndPluginManager()
...
+            'format_field' => 'SomeFormat',
+            'format_field_options' => 'SomeOptions',
+            'ckeditor_context' => null,
+            'ckeditor_basepath' => '',
+            'ckeditor_plugins' => array(),
+            'ckeditor_templates' => array(),
+            'ckeditor_toolbar_icons' => array(),
+        ));
+
+        $this->assertSame($view->vars['ckeditor_templates'], $templates);
```
Comment (ternel) [inline Tests/Form/Type/FormatterTypeTest.php]: Are the templates in the view's vars?

File: Tests/Form/Type/FormatterTypeTest.php (hunk 6 above)
Comment (ternel) [inline Tests/Form/Type/FormatterTypeTest.php]: Yep, but I can remove it... as you like

Comment (greg0ire) [inline Tests/Form/Type/FormatterTypeTest.php]: No no, it's fine, I just wasn't sure if you meant to add it.

File: Tests/Form/Type/FormatterTypeTest.php (hunk 7 above)
Comment (greg0ire) [inline Tests/Form/Type/FormatterTypeTest.php]: Are yo asking us a question or translating what this test does for us?

Comment (ternel) [inline Tests/Form/Type/FormatterTypeTest.php]: haha, yeah, just translating :)

**Comment by OskarStark** [comment]
Thank you, i changed the commit msg a bit and squashed them into one:

`Add ckeditor templates support for sonata_formatter_type and sonata_simple_formatter_type (#224)`

**Comment by ternel** [comment]
Thanks!
